node/src/components/linear_chain_sync.rs

```diff
@@ -117,18 +246,52 @@ impl<I: Clone + 'static> LinearChainSync<I> {
         REv: ReactorEventT<I>,
     {
         let peer = self.random_peer_unsafe(rng);
-        match self.linear_chain.pop() {
+
+        let curr_state = mem::replace(&mut self.state, State::None);
+
+        let (next_state, next_block) = match curr_state {
+            State::None | State::Done => {
+                panic!("Tried fetching next block when in {:?} state.", curr_state)
+            }
+            State::SyncingTrustedHash {
+                trusted_hash,
+                highest_block_seen,
+                mut linear_chain,
+                ..
+            } => {
+                let next_block = linear_chain.pop();
+                let new_state = State::SyncingTrustedHash {
+                    trusted_hash,
+                    highest_block_seen,
+                    linear_chain,
+                };
+                (new_state, next_block)
+            }
+            State::SyncingDescendants {
+                mut linear_chain,
+                trusted_hash,
+                highest_block_seen,
+            } => {
+                let next_block = linear_chain.pop_front();
+                let new_state = State::SyncingDescendants {
+                    linear_chain,
+                    trusted_hash,
+                    highest_block_seen,
+                };
+                (new_state, next_block)
+            }
+        };
+
+        self.state = next_state;
+
+        match next_block {
```

Comment:
```suggestion
        let maybe_next_block = match self.state {
            State::None | State::Done => {
                panic!("Tried fetching next block when in {:?} state.", self.state)
            }
            State::SyncingTrustedHash {
                ref mut linear_chain,
                ..
            } => {
                linear_chain.pop()
            }
            State::SyncingDescendants {
                ref mut linear_chain,
                ..
            } => {
                linear_chain.pop_front()
            }
        };

        match maybe_next_block {
```

Reply:
Right, I don't need to replace `self.state` anymore.

Reply:
Done.